### PR TITLE
[ENG-3685] Add permissions for withdrawn registration files

### PR DESCRIPTION
**Pull request by uditijmehta**
## Purpose

Implement custom permissions to handle file access for withdrawn registrations.

## Changes

- Added WithdrawnRegistrationPermission and updated FileDetailPermission to enforce access control for withdrawn registrations.
- Restrict API response

## Ticket

<(https://openscience.atlassian.net/browse/ENG-3685)>

### CODE REVIEW

File: api/files/permissions.py [hunk 1]
```diff
@@ -36,3 +38,41 @@ def has_object_permission(self, request, view, obj):
...
+        auth = get_user_auth(request)
+        if request.method in permissions.SAFE_METHODS:
+            return target.is_public or target.can_view(auth)
+
+        required_permission = self.REQUIRED_PERMISSIONS.get(request.method)
+        if required_permission:
+            return target.has_permission(auth.user, required_permission)
+        return True
+
+class FileDetailPermission(WithdrawnRegistrationPermission, permissions.BasePermission):
```
Comment (brianjgeiger) [inline api/files/permissions.py:77]: Under what circumstances might the WithdrawnRegistrationPermission be used on something that's not a File? It seems pretty file specific. We could probably just combine these two classes into one `RegistrationFileDetailPermission` instead.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
One minor thing, plus it might be nice if you could add in a couple of tests for this.

**Comment by jwalz** [comment]
So, I looked into this a while ago, and I think the actual easiest thing to do is to add a check for `target.is_retracted` the same way we do for `target.deleted` here (or, technically, `getattr(target, 'is_retracted', False)`): https://github.com/CenterForOpenScience/osf.io/blob/develop/api/files/views.py#L53

I don't love mixing permissions in with the `get_queryset` functionality, but a) we already do it and b) it feels significantly more appropriate for something that we want to universally mark `GONE` than for actual permissions.

Also, want to update `get_auth` [here](https://github.com/CenterForOpenScience/osf.io/blob/develop/addons/base/views.py#L305) to make sure that WB download links also break (since retractions are always visible)

**Review by brianjgeiger — APPROVED** [review]

File: api/files/permissions.py [hunk 2]
```diff
@@ -36,3 +38,38 @@ def has_object_permission(self, request, view, obj):
             return super(IsPreprintFile, self).has_object_permission(request, view, obj.target)
 
         return True
+
+
+class RegistrationFileDetailPermission(permissions.BasePermission):
```
Comment (jwalz) [inline api/files/permissions.py:43]: So, I don't think this is actually necessary with the change to `get_file`. Sorry for the misleading ticket. I meant to update it after my explorations a couple weeks ago but didn't notice that this ticket actually got scheduled. >_<

Should be sufficient to just comment out the inclusion of this permission and re-run tests locally to see if that's right.

File: api/files/serializers.py [hunk 3]
```diff
@@ -442,6 +442,16 @@ def to_representation(self, value):
         if guid:
             data['data']['id'] = guid._id
 
+        if isinstance(value.target, Registration) and value.target.is_retracted:
```
Comment (jwalz) [inline api/files/serializers.py:445]: Is this actually reachable? Shouldn't the view 410 before any data hits the serializer?

File: api_tests/files/views/test_file_detail.py [hunk 4]
```diff
@@ -705,6 +712,73 @@ def test_load_and_property(self, app, user, file):
...
+        retraction = resource.retract_registration(
+            user=resource.creator,
+            justification='Justification for retraction',
+            save=True,
+            moderator_initiated=False
+        )
+
+        approval_token = retraction.approval_state[resource.creator._id]['approval_token']
+        retraction.approve(user=resource.creator, token=approval_token)
+        resource.save()
```
Comment (jwalz) [inline api_tests/files/views/test_file_detail.py:726]: fun fact, we have an internal shorthand for this:
`retraction.accept()` with no Auth will skip the required approval steps.

File: api/files/views.py [hunk 5]
```diff
@@ -72,6 +76,7 @@ class FileDetail(JSONAPIBaseView, generics.RetrieveUpdateAPIView, FileMixin):
         CheckedOutOrAdmin,
         base_permissions.TokenHasScope,
         PermissionWithGetter(ContributorOrPublic, 'target'),
+        RegistrationFileDetailPermission,
```
Comment (jwalz) [inline api/files/views.py:79]: If the permission class *is* needed, then it's a pure replacement for the `PermissionWithGetter(...)` above, and that line can be removed.

**Review by jwalz — APPROVED** [review]

**Pull request merged**